Source/CombatExtended/CombatExtended/Comps/CompAmmosetSwitcher.cs

```diff
@@ -0,0 +1,262 @@
+
+        public CompProperties_FireModes compPropsFireModes => this.parent.def.comps.Find(x => x is CompProperties_FireModes) as CompProperties_FireModes;
+
+        public VerbProperties defVerbProps => this.parent.def.Verbs.Find(x => x is VerbPropertiesCE);
+
+        public CompProperties_AmmoUser compPropsAmmo => (CompProperties_AmmoUser)this.parent.def.comps.Find(x => x is CompProperties_AmmoUser);
+
+        public AmmoDef mainGunLoadedAmmo;
+
+        public int mainGunMagCount;
+
+        public AmmoDef UnderBarrelLoadedAmmo;
+
+        public int UnderBarrelMagCount;
+
+        public bool usingUnderBarrel;
+
+        public void SwitchToUB()
+        {
+            mainGunLoadedAmmo = compAmmo.CurrentAmmo;
+            mainGunMagCount = compAmmo.CurMagCount;
+
+            compAmmo.props = this.Props.propsUnderBarrel;
+            compEq.PrimaryVerb.verbProps = Props.verbPropsUnderBarrel;
+            compFireModes.props = this.Props.propsFireModesUnderBarrel;
+            compAmmo.CurMagCount = UnderBarrelMagCount;
+            compAmmo.CurrentAmmo = UnderBarrelLoadedAmmo;
+            compAmmo.SelectedAmmo = compAmmo.CurrentAmmo;
+            if (compAmmo.Wielder != null)
+            {
+                compAmmo.Wielder.TryGetComp<CompInventory>().UpdateInventory();
+
+                if (compAmmo.Wielder.jobs?.curJob?.def == CE_JobDefOf.ReloadWeapon)
+                {
+                    compAmmo.Wielder.jobs.EndCurrentJob(Verse.AI.JobCondition.InterruptForced);
+                }
+            }
+            compEq.PrimaryVerb.verbProps.burstShotCount = this.Props.propsFireModesUnderBarrel.aimedBurstShotCount;
+            usingUnderBarrel = true;
+        }
+
+        public void SwithToB()
+        {
+            UnderBarrelLoadedAmmo = compAmmo.CurrentAmmo;
+            UnderBarrelMagCount = compAmmo.CurMagCount;
+
+            compAmmo.props = compPropsAmmo;
+            compEq.PrimaryVerb.verbProps = defVerbProps.MemberwiseClone();
+            compFireModes.props = compPropsFireModes;
+            compAmmo.CurMagCount = mainGunMagCount;
+            compAmmo.CurrentAmmo = mainGunLoadedAmmo;
+            compAmmo.SelectedAmmo = compAmmo.CurrentAmmo;
+            if (compAmmo.Wielder != null)
+            {
+                compAmmo.Wielder.TryGetComp<CompInventory>().UpdateInventory();
+
+                if (compAmmo.Wielder.jobs?.curJob?.def == CE_JobDefOf.ReloadWeapon)
+                {
+                    compAmmo.Wielder.jobs.EndCurrentJob(Verse.AI.JobCondition.InterruptForced);
+                }
+            }
+            compEq.PrimaryVerb.verbProps.burstShotCount = defVerbProps.burstShotCount;
+            usingUnderBarrel = false;
+        }
+
+        public override IEnumerable<Gizmo> CompGetGizmosExtra()
+        {
+            if (compAmmo.Props.ammoSet == compPropsAmmo.ammoSet)
+            {
+                yield return new Command_Action

```

Comment:
This should also be cached.  When the pawn is selected, so gizmos are shown, this will run repeatedly, even with the time stopped.  If you had a pile of pawns with UB weapons, the cost would add up.